File: backend/src/main/java/com/just/donate/flow/DonationPart.java
package com.just.donate.flow;

import com.just.donate.utils.Splittable;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Optional;

import static com.just.donate.utils.Utils.*;

/**
 * Represents a part of a donation that can be spent on an expense. It is mainly used to keep track of the amount
 * of the donation that has been split up and spent on different expenses.
 */
public class DonationPart implements Splittable<DonationPart, BigDecimal>, java.util.Comparator<DonationPart> {

    private final BigDecimal amount;
    private final LocalDateTime donationTime;

    public DonationPart(BigDecimal amount, LocalDateTime donationTime) {
        this.amount = amount;
        this.donationTime = donationTime;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    @Override
    public Split<DonationPart, BigDecimal> splitOf(BigDecimal bigDecimal) {
        if (equal(bigDecimal, BigDecimal.ZERO)) {
            return new Split<>(Optional.empty(), Optional.of(this), Optional.empty());

        } else if (less(bigDecimal, amount)) {
            BigDecimal remainingAmount = amount.subtract(bigDecimal);
            return new Split<>(
                    Optional.of(new DonationPart(bigDecimal, donationTime)),
                    Optional.of(new DonationPart(remainingAmount, donationTime)),
                    Optional.empty());

        } else if (bigDecimal.equals(amount)) {
            return new Split<>(
                    Optional.of(this),
                    Optional.empty(),
                    Optional.empty());

        } else if (greater(bigDecimal, amount) && greater(amount, BigDecimal.ZERO)) {
            BigDecimal notPaid = bigDecimal.subtract(amount);
            return new Split<>(
                    Optional.of(this),
                    Optional.empty(),
                    Optional.of(notPaid));

        } else {
            throw new IllegalStateException("Should not happen?");
        }
    }

    @Override
    public String toString() {
        return amount.toString();
    }

    @Override
    public int compare(DonationPart o1, DonationPart o2) {
        return o1.donationTime.compareTo(o2.donationTime);
    }
}

File: backend/src/main/scala/com/just/donate/models/DonationPart.scala
package com.just.donate.models

import com.just.donate.models.Types.DonationGetter
import com.just.donate.utils.structs.{Split, Splittable}
import scala.math.Ordered.orderingToOrdered

import java.time.LocalDateTime
import com.just.donate.utils.Money
// TODO: when encoding and decoding json, make sure the connection between the donation and it's parts isn't lost
case class DonationPart(amount: Money, donationId: String, donationDate: LocalDateTime)
    extends Splittable[DonationPart, Money]
    with Ordering[DonationPart]:

  override def splitOf(split: Money): Split[DonationPart, Money] =
    if split == Money.ZERO then Split(None, Some(this), None)
    else if split < amount then
      Split(
        Some(DonationPart(split, donationId, donationDate)),
        Some(DonationPart(amount - split, donationId, donationDate)),
        None
      )
    else if split == amount then Split(Some(this), None, None)
    else if split > amount && amount > Money.ZERO then Split(Some(this), None, Some(split - amount))
    else throw new IllegalStateException("Should not happen?")

  override def toString: String = amount.toString

  override def compare(x: DonationPart, y: DonationPart): Int =
    x.donationDate.compareTo(y.donationDate)

  def donation(using donationGetter: DonationGetter): Option[Donation] =
    donationGetter(donationId)
  
  def earmarking(using donationGetter: DonationGetter): Option[String] =
    donationGetter(donationId).flatMap(_.earmarking)